### PR TITLE
Remove `normalize-path` package

### PR DESCRIPTION
**Pull request by fisker**
- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory)
- [ ] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/pr-XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by thorn0** [comment]
This breaks in the browser:
```console
> prettier.format("a < b > ( c )", { plugins: prettierPlugins, filepath: 'foo.ts' });
19:46:09.835 standalone.js:13126 Uncaught TypeError: path.basename is not a function
    at inferParser (standalone.js:13126)
    at normalize$1 (standalone.js:13028)
    at formatWithCursor (standalone.js:16704)
    at standalone.js:31935
    at Object.format (standalone.js:31944)
    at <anonymous>:1:10
inferParser @ standalone.js:13126
normalize$1 @ standalone.js:13028
formatWithCursor @ standalone.js:16704
(anonymous) @ standalone.js:31935
format @ standalone.js:31944
(anonymous) @ VM36:1
```

**Comment by fisker** [comment]
@thorn0 how to do this test ?

**Comment by fisker** [comment]
Oh. script include standalone.js ?

**Comment by thorn0** [comment]
https://jsbin.com/giqiroy/edit?html,js,console

**Comment by fisker** [comment]
I think `filepath.split(/[/\\]/).pop()` should work.

**Comment by thorn0** [comment]
That should work. I wonder why the tests didn't catch this. There is an Azure job configured to run the tests with `TEST_STANDALONE=true`, but it didn't catch this.

**Comment by fisker** [comment]
I'll figure

**Comment by thorn0** [comment]
I found why: `parser` is passed explicitly in all those tests.

**Comment by fisker** [comment]
@thorn0 `STANDALONE` only test `tests/`
 https://github.com/prettier/prettier/blob/6a497b691440dd6ce8b713fec1326a2b6a23e3e6/scripts/test-dist.js#L24

**Comment by thorn0** [comment]
Right. And `parser` is explicitly specified in all those tests.

**Comment by fisker** [comment]
This function is `inferParser`, and all tests in `tests/` has options `parser`. so we don't have to look for parser, right?

**Comment by fisker** [comment]
I'm trying to add a `no-parser` test

**Comment by fisker** [comment]
Let's see if https://github.com/prettier/prettier/pull/7465/commits/7a9fb9fc46a03b451c6ffb10e6e219e881551f89 can catch this error

**Comment by fisker** [comment]
Error caught https://dev.azure.com/prettier/prettier/_build/results?buildId=4915&view=logs&j=cc0821cc-6ec9-5a14-8bb9-4961b231143e&t=0577ecdd-8efd-50b6-992b-f5e447724f2c&l=1959

File: tests_config/run_spec.js [hunk 1]
```diff
@@ -15,8 +15,11 @@ const prettier = !TEST_STANDALONE
...
 
 global.run_spec = (dirname, parsers, options) => {
-  // istanbul ignore next
-  if (!parsers || !parsers.length) {
+  // `isNoParserTests` mean to test `inferParser` on `standalone`
+  const isNoParserTests = dirname.includes("no-parser-tests");
+  if (isNoParserTests) {
+    parsers = [];
+  } else if (!parsers || !parsers.length) {
     throw new Error(`No parsers were specified for ${dirname}`);
```
Comment (thorn0) [inline tests_config/run_spec.js]: Is this check really needed? I think we can remove it.

Comment (fisker) [reply tests_config/run_spec.js]: Maybe new tests forgot add parsers, I think we should keep it

Comment (fisker) [reply tests_config/run_spec.js]: Added in https://github.com/prettier/prettier/pull/3356 , let's keep it. but `// istanbul ignore next` can be removed...

File: scripts/build/shims/path.js [hunk 2]
```diff
@@ -1,9 +1,16 @@
...
 export function extname(path) {
+  path = path.split(sep).pop();
   const dotIndex = path.lastIndexOf(".");
   if (dotIndex === -1) return "";
   return path.slice(dotIndex);
 }
 
+export function basename(path) {
+  return path.split(sep).pop();
+}
```
Comment (fisker) [inline scripts/build/shims/path.js]: I'm going to fix with this shim.

File: scripts/build/shims/path.js [hunk 3]
```diff
@@ -1,9 +1,16 @@
+const sep = /[\\/]/;
+
 export function extname(path) {
+  path = path.split(sep).pop();
```
Comment (fisker) [inline scripts/build/shims/path.js]: this fixes `/a.b/c`

File: tests_config/run_spec.js (hunk 1 above)
Comment (thorn0) [reply tests_config/run_spec.js]: What's wrong with allowing Prettier to infer the parser if it's what the author of the test wants? I guess this check was added when the parser inference wasn't as good as it is today.

Comment (fisker) [reply tests_config/run_spec.js]: I guess, if someone want test 'foo.js' on `flow` but forgot add parser, the default parser `babel` will work, and a little hard to find the mistake?

File: tests_config/run_spec.js [hunk 4]
```diff
@@ -15,8 +15,11 @@ const prettier = !TEST_STANDALONE
   : require("prettier/standalone");
 
 global.run_spec = (dirname, parsers, options) => {
-  // istanbul ignore next
-  if (!parsers || !parsers.length) {
+  // `isNoParserTests` mean to test `inferParser` on `standalone`
+  const isNoParserTests = dirname.includes("no-parser-tests");
```
Comment (thorn0) [inline tests_config/run_spec.js]: Let's rename the folder to `parser-inference` to better describe what is actually tested.

Comment (fisker) [reply tests_config/run_spec.js]: sure

**Comment by fisker** [comment]
> I see `eval("require").resolve` in multiple places, could we factor this out somehow?

Suggestion?

**Comment by brodybits** [comment]
> > I see `eval("require").resolve` in multiple places, could we factor this out somehow?
> 
> Suggestion?

<https://github.com/fisker/prettier/pull/300>

A side point is that we could do something similar for eval("require")(...) in `src/common/internal-plugins.js` and `src/cli/util.js`.

**Comment by fisker** [comment]
@brodybits if we want do this, maybe better `const xx = eval('require')` some where

See https://github.com/prettier/prettier/blob/master/src/common/internal-plugins.js

**Comment by brodybits** [comment]
> `const xx = eval('require')` some where
> 
> See https://github.com/prettier/prettier/blob/master/src/common/internal-plugins.js

Yes. I am starting to wonder if we should do this in a separate (followup) PR and factor more eval calls out of that module at the same time?

**Comment by fisker** [comment]
I'm not suggesting to do this way, I think `eval('require')` is good, can't find a better solution, as we may don't need it someday. Also `eval('require').resolve` should be good.

**Comment by fisker** [comment]
> > `const xx = eval('require')` some where
> > See https://github.com/prettier/prettier/blob/master/src/common/internal-plugins.js [](/prettier/prettier/blob/HEAD@{2020-02-02T12:16:17Z}/src/common/internal-plugins.js)
> 
> Yes. I am starting to wonder if we should do this in a separate (followup) PR and factor more eval calls out of that module at the same time?

I don't think we should do anything, it's a hack. It works just like `require/*please-keep-me-as-it-is*/(foo)`

**Comment by brodybits** [comment]
Please disregard my comments on eval, they were intended for another PR. I will raise a new PR to propose what I would like to do.

**Pull request merged**